djvm/src/main/kotlin/net/corda/djvm/source/SourceClassLoader.kt

```diff
@@ -22,45 +23,80 @@ import java.nio.file.Paths
 import kotlin.streams.toList
 
 /**
- * Class loader to manage an optional JAR of replacement Java APIs.
+ * Base interface for API and user sources.
  */
-class BootstrapClassLoader private constructor(
-    jarPaths: List<Path>
-) : URLClassLoader(resolvePaths(jarPaths), null) {
-    /**
-     * @param bootstrapJar The location of the JAR containing the Java APIs.
-     */
-    constructor(bootstrapJar: Path) : this(listOf(bootstrapJar))
-    constructor() : this(emptyList())
+interface Source : Closeable {
+    val loadResource: (String) -> URL?
+}
+
+/**
+ * This is the contract that our source of Java APIs must satisfy.
+ */
+interface ApiSource : Source
 
+/**
+ * This is the contract that the source of the user's own code must satisfy.
+ * It will almost certainly be a [ClassLoader] of some description.
+ */
+interface UserSource : Source {
+    val loadClass: (String) -> Class<*>
+    fun getURLs(): Array<URL>
```

Comment:
Out of curiosity, why is `loadClass` a lambda value rather than a function?

Reply:
Mainly for consistency with the `loadResource` lambda, which was a "clever" way to side-step the fact that `ClassLoader.findResource` is `protected` rather than `public` and that neither `ApiSource` nor `UserSource` needed to be a `ClassLoader`.